tomviz/vtkChartHistogramColorOpacityEditor.cxx

```diff
@@ -0,0 +1,262 @@
+}
+
+vtkAxis* vtkChartHistogramColorOpacityEditor::GetHistogramAxis(int axis)
+{
+  return this->HistogramChart->GetAxis(axis);
+}
+
+bool vtkChartHistogramColorOpacityEditor::GetCurrentControlPointColor(double rgb[3])
+{
+  vtkColorTransferFunction* ctf =
+    this->ColorTransferControlPointsItem->GetColorTransferFunction();
+  if (!ctf)
+  {
+    return false;
+  }
+
+  vtkIdType currentIdx = this->ColorTransferControlPointsItem->GetCurrentPoint();
+  if (currentIdx < 0)
+  {
+    return false;
+  }
+
+  double xrgbms[6];
+  ctf->GetNodeValue(currentIdx, xrgbms);
+  rgb[0] = xrgbms[1];
+  rgb[1] = xrgbms[2];
+  rgb[2] = xrgbms[3];
+
+  return true;
+}
+
+void vtkChartHistogramColorOpacityEditor::SetCurrentControlPointColor(const double rgb[3])
+{
+  vtkColorTransferFunction* ctf =
+    this->ColorTransferControlPointsItem->GetColorTransferFunction();
+  if (!ctf)
+  {
+    return;
+  }
+
+  vtkIdType currentIdx = this->ColorTransferControlPointsItem->GetCurrentPoint();
+  if (currentIdx < 0)
+  {
+    return;
+  }
+
+  double xrgbms[6];
+  ctf->GetNodeValue(currentIdx, xrgbms);
+  xrgbms[1] = rgb[0];
+  xrgbms[2] = rgb[1];
+  xrgbms[3] = rgb[2];
+  ctf->SetNodeValue(currentIdx, xrgbms);
+}
+
+double vtkChartHistogramColorOpacityEditor::GetContourValue()
+{
+  return this->HistogramChart->GetContourValue();
+}
+
+bool vtkChartHistogramColorOpacityEditor::Paint(vtkContext2D* painter)
+{
+  vtkContextScene* scene = this->GetScene();
+  int sceneWidth = scene->GetSceneWidth();
+  int sceneHeight = scene->GetSceneHeight();
+  if (this->Private->NeedsUpdate ||
+      sceneWidth != this->Private->Geometry.GetX() ||
+      sceneHeight != this->Private->Geometry.GetY())
+  {
+    this->Private->NeedsUpdate = false;
+
+    // Update the geometry size cache
+    this->Private->Geometry.Set(sceneWidth, sceneHeight);
+
+    // Upper chart (histogram) expands, lower chart (color bar) is fixed height.
+    float x = this->Borders[vtkAxis::LEFT];
+    float y = this->Borders[vtkAxis::BOTTOM];
+
+    // Add the width of the left axis to x to make room for y labels
+    this->GetHistogramAxis(vtkAxis::LEFT)->Update();
+    float leftAxisWidth = this->GetHistogramAxis(vtkAxis::LEFT)->
+      GetBoundingRect(painter).GetWidth();
+    x += leftAxisWidth;
+
+    float colorBarThickness = 20;
+    float plotWidth = sceneWidth - x - this->Borders[vtkAxis::RIGHT];
+
+    vtkRectf colorTransferFunctionChartSize(x, y, plotWidth, colorBarThickness);
+    this->ColorTransferFunctionChart->SetSize(colorTransferFunctionChartSize);
+
+    float bottomAxisHeight = this->GetHistogramAxis(vtkAxis::BOTTOM)->
+      GetBoundingRect(painter).GetHeight();
+    float verticalMargin = bottomAxisHeight;
+    y += colorBarThickness + verticalMargin - 5;
+    vtkRectf histogramChart(x, y, plotWidth, sceneHeight - y - this->Borders[vtkAxis::TOP]);
+    this->HistogramChart->SetSize(histogramChart);
+  }
+
+  return this->Superclass::Paint(painter);
+}
```

Comment:
Thanks for catching. I'm trying out Sublime Text 3 and newline insertion is off by default. Fixed now.